CV-310: Horizon training with non-square input size

Zwicii opened the pull request:
## What?
 
Add support for horizon training when the input size is not square.

 
## Why?
 
We want to also train horizon model with not square inputs.
 
## How?

> [!WARNING]  
> Because I had to do a lot of testing, the commit history ended up being a mess. I squashed the commits to clean it, but since I'm not always the brightest star in the universe, I forgot to force push. As a result, we now have both the individual commits and the squashed ones 🫠🫠🫠.
> **I have linked the commits that contain the actual changes to make it a bit better 🙃.**
 
### Handling Input Size

If `imgsz` is provided as a single integer, it will be converted into a `tuple` with equal width and height.
This conversion is necessary for the `horizon.train.run(**train_args)` and also for the  `check_train_batch_size()` function, as other training codes also rely on this logic (but maybe there is a better way 💡)

**Example**

```python
# Ensure imgsz is a tuple (height, width)
imgsz = (imgsz, imgsz) if isinstance(imgsz, int) else imgsz
```

### Modifying the Horizon Dataloader for Non-Square Inputs
Instead of assuming square dimensions, explicitly set `height` and `width`.

**Example (Assumes Square Input)**

```python
A.PadIfNeeded(
    min_height=imgsz, min_width=imgsz, border_mode=cv2.BORDER_CONSTANT, value=0
)
```

**After (Handles Rectangular Inputs)**
```python
A.PadIfNeeded(
    min_height=imgsz[0], min_width=imgsz[1], border_mode=cv2.BORDER_CONSTANT, value=0
)
```

> [!CAUTION]
> **Uncertain Part: A.ResizeIfNeeded(max_size=max(imgsz))**
> I used `max(imgsz)` because chatty said: Use `max(imgsz)` if you want to resize while ensuring neither dimension exceeds the limit. This keeps the aspect ratio intact and ensures the image does not become larger than the specified max size.
> **but not sure about this**

### Additional Bugfixes  

- If the batch size was not `640`, it would always default to `16`.  
  - This was fixed by removing the condition. 
  - [commit](https://github.com/SEA-AI/yolov5/commit/01843c414b1e9dfbb666a696431b8675850142cd)

- CUDA device selection was not working because the device was hardcoded to `cuda:0`.  
  - Now, the selected device is properly returned in the `select_device()` method.  
  - [commit](https://github.com/SEA-AI/yolov5/commit/2c42e5a34debfea922a3f18088be9816e72061e7)

- Issue with `os.environ["CUDA_VISIBLE_DEVICES"] = device`:  
  - If a GPU with `idx=1` was selected, it was set as the only visible GPU.  
  - However, when using automatic GPU selection, accessing device properties via:  
    ```python
    properties = torch.cuda.get_device_properties(device)  # device properties
    ```  
    would fail because **if only one GPU is visible, its index becomes `0`, not `1`**.  
  - This was fixed by always exposing all GPUs up to the selected index, preventing issues with reordered indices.  
    - For example, if `gpu=1` is selected, both `0` and `1` are exposed.
    - [commit](https://github.com/SEA-AI/yolov5/commit/21f5368304dd023de57293a5bdbde744776d3de5)

- Auto batch size calculation was incorrect because GPU `0` was always used.  
  - Now, the correct device is passed, ensuring accurate memory calculations:  
    ```python
    mem = torch.cuda.memory_reserved(device) / 1e9 if torch.cuda.is_available() else 0  # (GB)
    ```
    - [commit](https://github.com/SEA-AI/yolov5/commit/21f5368304dd023de57293a5bdbde744776d3de5)




## Backout plan

Revert this changes.

 
## Testing



 
A training was started with size `(960x1280)`.
[W&B run](https://wandb.ai/sea-ai/yolo-horizon/runs/rh0kjf76?nw=nwuserseaaimlops)

Zwicii committed:
CV-310: make horizon training work with height and width imgsz

Zwicii committed:
CV-310: try to fix issue

Zwicii committed:
CV-310: update batch size handling in get_dataloaders

Zwicii committed:
CV-310: update imgsz parameter handling to make it more robust

Zwicii committed:
CV-310: fix issue regarding cuda device selection

Zwicii committed:
CV-310: add device setup function to configure CUDA settings

Zwicii committed:
add debug message

Zwicii committed:
add debug messages

Zwicii committed:
try to fix bug

Zwicii committed:
last try

Zwicii committed:
final last try

Zwicii committed:
try to fix problem

Zwicii committed:
try to fix bug

Zwicii committed:
try to fix bug

Zwicii committed:
try to fix bug

Zwicii committed:
fix bug

Zwicii committed:
CV-310: fix issues regarding gpu selection

Zwicii committed:
CV-310: remove debug messages

Zwicii committed:
Merge branch 'CV-310-imgsz-width-height' of https://github.com/SEA-AI/yolov5 into CV-310-imgsz-width-height

Zwicii committed:
CV-310: fix problem regarding auto batch size calculation

Zwicii committed:
add debug messages for autobatch

Zwicii committed:
try to fix issue

Zwicii committed:
try to fix issue

Zwicii committed:
debug problem

Zwicii committed:
test

Zwicii committed:
add debug messages

Zwicii committed:
set all devices visible

Zwicii committed:
test

Zwicii committed:
test

Zwicii committed:
try to fix bug

Zwicii committed:
CV-310: fix problem regarding auto batch size calculation

Zwicii committed:
Merge branch 'CV-310-imgsz-width-height' of https://github.com/SEA-AI/yolov5 into CV-310-imgsz-width-height

Zwicii committed:
CV-310: ensure consistent syntax

Zwicii committed:
CV-310: update resizing logic in geometric transformations to use max size

Zwicii committed:
Revert "CV-310: update resizing logic in geometric transformations to use max size"

This reverts commit 46d0e3e927a8e641274d5c354006d83d9a08f73a.

Zwicii committed:
Merge branch 'develop' into CV-310-imgsz-width-height

kevinconka commented on horizon/transforms.py:
```diff
@@ -29,16 +30,16 @@ def log_transforms(transforms: List[BasicTransform], prefix: str) -> None:
     return [
         A.RandomCropFromBorders(p=0.1),
-        Ax.ResizeIfNeeded(max_size=imgsz),
+        Ax.ResizeIfNeeded(max_size=min(imgsz)),
```
This should be `max(imgsz)`

Image you have a 720x1280 image and your CNN input is 640x960
- If we pass  `min(640, 960)` then it will be resized to 360x640 🙅 
- If we pass `max(640, 960)` then it will be resized to 540x960 👌 

I was also hesitant, so I tried this code:

```python
import cv2
import utils.albumextensions as Ax
import albumentations as A
from PIL import Image

imgsz = (640, 960)
T = A.Compose(
    [
        A.RandomCropFromBorders(p=0.1),
        Ax.ResizeIfNeeded(max_size=min(imgsz)),
        A.HorizontalFlip(p=0.5),
        A.PadIfNeeded(min_height=imgsz[0], min_width=imgsz[1], border_mode=cv2.BORDER_CONSTANT, value=0),
    ]
)

im = cv2.cvtColor(cv2.imread("/Users/kevinserrano/Downloads/18095390.jpg"), cv2.COLOR_BGR2RGB)
print(im.shape)  # (720, 1280, 3)
t_im = T(image=im)["image"]
print(t_im.shape)  # (640, 960, 3)
Image.fromarray(t_im)
```

![image](https://github.com/user-attachments/assets/320b900e-d196-45a5-9543-82f0f196ec8c)

If I use `Ax.ResizeIfNeeded(max_size=max(imgsz))`

![image](https://github.com/user-attachments/assets/301ce74d-778e-41b8-b2fc-fae67bcb4da4)


The intention behind this custom augmentation is basically to only downscale but never upscale.
I was not able to find any out-of-the-box augmentation to do it back then. Maybe a better name would be `DownscaleOnly(max_size=...)` 😛

Zwicii replied:
ahhhhhhhhhhhhh 🫠
i need to check this - using max size resulted in an error - but makes sense should have tried it or mentioned it soooorry

kevinconka replied:
so... found out why it is failing....

An easy example is:
- CNN input is 640x960 --> `Ax.ResizeIfNeeded(960)`
- Image size is 960x720 --> No resize needed since `960 == max(im.shape)`, but... this is wrong because 720 > 640 ☠️ 

```
imgsz = (640, 960)
T = A.Compose(
    [
        A.RandomCropFromBorders(p=0.1),
        Ax.ResizeIfNeeded(max_size=max(imgsz)),
        A.HorizontalFlip(p=0.5),
        A.PadIfNeeded(min_height=imgsz[0], min_width=imgsz[1], border_mode=cv2.BORDER_CONSTANT, value=0),
    ]
)

im = cv2.cvtColor(cv2.imread("/Users/kevinserrano/Downloads/18095390.jpg"), cv2.COLOR_BGR2RGB)
im = cv2.resize(im, (960, 720))
print(im.shape)  # (720, 960, 3)
t_im = T(image=im)["image"]
print(t_im.shape)  # (640, 960, 3) != (720, 960, 3)
assert imgsz == t_im.shape[:2], f"imgsz={imgsz} != t_im.shape[:2]={t_im.shape[:2]}"
Image.fromarray(t_im)
```

```console
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[6], [line 24](vscode-notebook-cell:?execution_count=6&line=24)
     [22] t_im = T(image=im)["image"]
     [23] print(t_im.shape)  # (640, 960, 3) != (720, 960, 3)
---> [24] assert imgsz == t_im.shape[:2], f"imgsz={imgsz} != t_im.shape[:2]={t_im.shape[:2]}"
     [25] Image.fromarray(t_im)

AssertionError: imgsz=(640, 960) != t_im.shape[:2]=(720, 960)
```

kevinconka replied:
> [!NOTE]
> Fxed by adding support for specifying max size for both dims. This was done by introducing the arg `max_size_hw` similar to albumentation's latest implementations: https://arc.net/l/quote/snrgksxl

After some agent-assisted changes, went from this

<img width="1058" alt="image" src="https://github.com/user-attachments/assets/a4e0352c-07c4-4487-a8d9-38fc32b2799e" />

to this

<img width="1058" alt="image" src="https://github.com/user-attachments/assets/7485118c-8c60-452f-9b9f-cc4e9c1e013b" />

Zwicii committed:
Merge branch 'develop' into CV-310-imgsz-width-height

kevinconka commented on utils/autobatch.py:
```diff
@@ -48,11 +53,12 @@ def autobatch(model, imgsz=640, fraction=0.8, batch_size=16):
     batch_sizes = [1, 2, 4, 8, 16]
     try:
-        img = [torch.empty(b, 3, imgsz, imgsz) for b in batch_sizes]
+        img = [torch.empty(b, 3, imgsz[0], imgsz[1]) for b in batch_sizes]
```
oohhh nice!

kevinconka commented on utils/autobatch.py:
```diff
@@ -10,13 +10,18 @@
 
 
-def check_train_batch_size(model, imgsz=640, amp=True):
+def check_train_batch_size(model, imgsz=[640, 640], amp=True):
```
I don't think we need to change the default arg, might break some other stuff in normal yolo.

kevinconka commented on utils/autobatch.py:
```diff
@@ -10,13 +10,18 @@
 
-
-def autobatch(model, imgsz=640, fraction=0.8, batch_size=16):
+def autobatch(model, imgsz=[640, 640], fraction=0.8, batch_size=16):
```
I would move the `imgsz = (imgsz, imgsz) if isinstance(imgsz, int) else imgsz` here, since it is only used in this function.

kevinconka commented on utils/autobatch.py:
```diff
@@ -10,13 +10,18 @@
     """Checks and computes optimal training batch size for YOLOv5 model, given image size and AMP setting."""
+
+    # check if imgsize is a single number and convert it to tuple otherwise
+    imgsz = (imgsz, imgsz) if isinstance(imgsz, int) else imgsz
```
move to `autobatch`

kevinconka commented on utils/torch_utils.py:
hmm... we would need to monitor regular yolo trainings to see if this does not break anything.

When you apply `os.environ["CUDA_VISIBLE_DEVICES"] = 1`, then you only have one device and therefore `"cuda:0"` is OK.

In general I don't see a major flaw, just worried that the device-always-zero _bug_ only happened with the horizon training and NOT with BBox yolo trainings 🤷

Zwicii replied:
yees thats super strange - i need to double check it

kevinconka commented on horizon/train.py:
It seems that the imgsz is passed around too much haha, I would like to be consistent and always have the `int | Tuple[int, int]` support in every function, but... perhaps it is a YAGNI or overkill 🤷 .

No need to change anything here though, just an observation.

kevinconka reviewed (changes requested):
If we launched a training with a rectangular shape, then... there might be a lot of excessive black padding because of the `Ax.ResizeIfNeeded(min(imgsz))` transform. So... we might need to re-launch it 🫣

Zwicii committed:
CV-310: refactor autobatch

kevinconka committed:
Ax.ResizeIfNeeded(max_size=max(imgsz))

Signed-off-by: Kevin Serrano <35779409+kevinconka@users.noreply.github.com>

kevinconka committed:
Refactor ResizeIfNeeded to support max_size_hw parameter for image resizing

- Updated the ResizeIfNeeded class to accept max_size_hw as a tuple for maximum height and width constraints.
- Modified geometric_augment and horizon_base functions to utilize max_size_hw instead of max_size.
- Added validation for max_size_hw in the initialization schema to ensure correct input format.

kevinconka committed:
Fix slope_intercept_to_points since previous version assumed `w==h` and lead to wrong visualisation when using a rectangular CNN input shape.

kevinconka committed:
Add num_workers parameter to get_dataloaders in horizon/train.py

- Introduced num_workers argument to improve data loading efficiency.
- Simplified bolean logic suggested by linter

kevinconka committed:
Refactor get_wb_images and remove_black_padding functions in horizon/train.py to plot RGB images instead of only grayscale

- Updated image processing in get_wb_images to handle image dimensions correctly.
- Improved comments for clarity regarding background and horizon masks.
- Enhanced remove_black_padding to convert images to grayscale before thresholding for better contour detection.

kevinconka committed:
Fix mask wandb visualisation

kevinconka committed:
Fix y_mask initialization in get_wb_images to correctly create a 2D mask for horizon visualization

Zwicii committed:
Merge branch 'CV-310-imgsz-width-height' of https://github.com/SEA-AI/yolov5 into CV-310-imgsz-width-height

Zwicii committed:
CV-310: Refactor validation logic to make it better readable

kevinconka reviewed (dismissed):
<img src="https://media4.giphy.com/media/v1.Y2lkPWJkM2VhNTdlOGpmazQ5NHdsd2hhbHlxZWdodnRtYnlrMHo1cjRpcjQ5dzhhdGxjciZlcD12MV9naWZzX3RyZW5kaW5nJmN0PWc/1aLsEE397a70GrmXuH/giphy.gif"/>

Zwicii committed:
CV-310: revert changes in cuda device selection

Zwicii committed:
add AMP check to the batch size estimation in horizon/train.py

Zwicii committed:
fix import

Zwicii committed:
fix problems with GradScaler

Zwicii committed:
Update device parameter in model loading for improved flexibility

kevinconka commented on horizon/train.py:
```diff
@@ -309,7 +316,8 @@ def run(
     # Batch size
     if batch_size == -1:  # single-GPU only, estimate best batch size
-        batch_size = check_train_batch_size(model, imgsz)
+        amp = check_amp(model)  # check AMP
```
out of curiosity, why is this needed?

Zwicii replied:
> out of curiosity, why is this needed?

Its used for the batchsize calculation and it is default true. but for the horizon model its false and this was causing issues

kevinconka commented on models/custom.py:
```diff
@@ -50,7 +50,7 @@ def __init__(
 
         if Path(weights).is_file() or weights.endswith(".pt"):
-            model = attempt_load(weights, device="cpu", fuse=fuse)
+            model = attempt_load(weights, device=device, fuse=fuse)
```
hmm... nice, does this improve something? I copy+pasted the code from here https://github.com/SEA-AI/yolov5/blob/bc2bfc11c0dd9b4c5bbc2bab6de6eebe87fdd6d5/classify/train.py#L144

Zwicii replied:
no had no effect - 

> hmm... nice, does this improve something? I copy+pasted the code from here
> 
> https://github.com/SEA-AI/yolov5/blob/bc2bfc11c0dd9b4c5bbc2bab6de6eebe87fdd6d5/classify/train.py#L144

no had no effect - but I was thinking that we should pass the "used" device instead of hard code it

kevinconka reviewed (commented):
LGTM, can we launch two trainings to test? Then we can approve and merge

kevinconka committed:
Merge branch 'develop' into CV-310-imgsz-width-height

Zwicii committed:
CV-310: Remove duplicated logging of training batch size check in autobatch.py

Zwicii committed:
Merge branch 'CV-310-imgsz-width-height' of https://github.com/SEA-AI/yolov5 into CV-310-imgsz-width-height

Zwicii committed:
CV-310: revert  model loading to use CPU device again

Zwicii committed:
CV-310: revert memory reserved calculation in profile function by removing device parameter

kevinconka reviewed (approved):
<img src="https://media2.giphy.com/media/v1.Y2lkPWJkM2VhNTdldmN1djhsZDh3M3lrdWVuMHQzbXYxM3Q1bWhoMjBkMnZta2pmczg5dSZlcD12MV9naWZzX3NlYXJjaCZjdD1n/dnCNI8LEf7K4otcMqS/giphy.gif"/>
